engine/src/flutter/lib/web_ui/lib/src/engine/semantics/landmarks.dart

```diff
@@ -0,0 +1,187 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'label_and_value.dart';
+import 'semantics.dart';
+
+/// Indicates a complementary element.
+///
+/// Uses aria complementary role to convey this semantic information to the element.
+///
+/// Screen-readers takes advantage of "aria-label" to describe the visual.
+class SemanticComplementary extends SemanticRole {
+  SemanticComplementary(SemanticsObject semanticsObject)
+    : super.withBasics(
+        EngineSemanticsRole.complementary,
+        semanticsObject,
+        preferredLabelRepresentation: LabelRepresentation.ariaLabel,
+      ) {
+    setAriaRole('complementary');
+  }
+
+  @override
+  void update() {
+    super.update();
+
+    final Set<int> complementaryIds = _sameRoleIds(semanticsObject);
+
+    if (complementaryIds.length > 1) {
+      _updateUniqueLabels(complementaryIds, semanticsObject);
+    } else if (semanticsObject.label != null) {
+      setAttribute('aria-label', semanticsObject.label!);
+    }
+  }
+
+  @override
+  bool focusAsRouteDefault() => focusable?.focusAsRouteDefault() ?? false;
+}
+
+Set<int> _sameRoleIds(SemanticsObject semanticsObject) {
+  final Map<int, SemanticsObject> tree = semanticsObject.owner.semanticsTree;
+  final Set<int> sameRoleIds = {};
+  for (final int id in tree.keys) {
+    if (tree[id]?.role == semanticsObject.role) {
+      sameRoleIds.add(id);
+    }
+  }
+  return sameRoleIds;
+}
+
+void _updateUniqueLabels(Set<int> ids, SemanticsObject semanticsObject) {
+  final Map<int, SemanticsObject> tree = semanticsObject.owner.semanticsTree;
+  for (final int id in ids) {
+    String label = tree[id]?.label ?? '';
+    if (label == '') {
+      label = 'flt-semantic-node-$id';
```

Comment:
I am confused, why we use this as label? If a label is required for a certain role, we should enforce that from the framework side.

Reply:
ideally in the semantics role check, and we can assume the semantics node has a label in the web engine

Reply:
Only `region` requires a label, so I added the requirement from framework side.

Since I saw we should assign each of them a unique label when the same role is assigned to multiple places, so I thought maybe the node id can be the default value. Updating this from framework side!

Reply:
Yes, makes sense to me! Thanks a lot for your suggestions! I just updated the PR based on the comment:)